### PR TITLE
Fix wrong display of multiline messages in the log after filtering

### PR DESCRIPTION
**Pull request by jason810496**
closes: #41265

### Issue Context
After research, the issue is caused by `airflow/www/static/js/dag/details/taskInstance/Logs/utils.ts`, the API just return the full raw logs.
Additionally, incorrect display behavior also occurs when filtering by file source.

### Main Fix Idea
The `currentLogLevel` and `currentFileSource` should remain same until a new `logLevel` or `fileSource` is encountered.

### Screenshot After Fix

**Without any filters**
Without any filters
**Filter with level**
Filter with level
**Filter with file source**
Filter with file source
**Filter with level and file source**
Filter with level and file source

### CODE REVIEW

**Comment by jason810496** [comment]
Hi @Lee-W, I think this issue should be back ported to `2.9.x`, `2.8.x`, and `2.7.x` and also fixed on the latest `main` branch once it is merged. Does that sound correct?

**Comment by pierrejeambrun** [comment]
> Hi @Lee-W, I think this issue should be back ported to 2.9.x, 2.8.x, and 2.7.x

We do not backport to released versions. (unless critical security fixes but I haven't seen that happen before).

This will most likely be in the next patch release of `2.10.x`, people will have to upgrade to get it.

**Comment by Lee-W** [comment]
Thanks @pierrejeambrun ! 

@jason810496 To create a "backport" PR to 2.10.x, you'll need to branch out from `v2-10-test`. In case you've not yet tried it 🙂

**Comment by potiuk** [comment]
> @jason810496 To create a "backport" PR to 2.10.x, you'll need to branch out from `v2-10-test`. In case you've not yet tried it 🙂

Actually we now have cherry-picker - and we marked it with "backport" label, so the backport PR should be created automatically and if it is not possible, instructions what to do will be posted automatically here as comment :D

**Comment by pierrejeambrun** [comment]
Main is too different now and do not have this issue because of structured logs. (json)

I noticed that the PR directly targets `v2-10-test`. I think it's fine in this case.

@potiuk @Lee-W are we fine with that ?

File: airflow/www/static/js/dag/details/taskInstance/Logs/utils.test.tsx [hunk 1]
```diff
@@ -40,6 +40,7 @@ const mockTaskLog = `
 [2022-06-04 00:00:02,010] {taskinstance.py:1548} WARNING - Exporting env vars: AIRFLOW_CTX_DAG_OWNER=*** AIRFLOW_CTX_DAG_ID=test_ui_grid
 [2024-07-01 00:00:02,010] {taskinstance.py:1548} INFO - Url parsing test => "https://apple.com", "https://google.com", https://something.logs/_dashboard/?_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:now-1d,to:now))&_a=(columns:!(_source),filters:!(('$state':(store:appState))))
 `;
+const mockExtraLog = `--------------------------------------------------------------------------------`;
```
Comment (pierrejeambrun) [inline airflow/www/static/js/dag/details/taskInstance/Logs/utils.test.tsx]: Can we add an example with a stack trace for example that has intermediate lines, (typical case where only the first line is considered ERROR, while the entire trace should be).

I don't understand what this `mockExtraLog` is supposed to do.

File: airflow/www/static/js/dag/details/taskInstance/Logs/utils.test.tsx [hunk 2]
```diff
@@ -145,7 +156,8 @@ describe("Test Logs Utils.", () => {
       []
     );
 
-    const lines = parsedLogs!.split("\n");
+    // remove the last line which is empty
+    const lines = parsedLogs!.split("\n").filter((line) => line.length > 0);
```
Comment (pierrejeambrun) [inline airflow/www/static/js/dag/details/taskInstance/Logs/utils.test.tsx:202]: Why do we need that now ?

File: airflow/www/static/js/dag/details/taskInstance/Logs/utils.ts [hunk 3]
```diff
@@ -87,24 +89,19 @@ export const parseLogs = (
   // Coloring (blue-60 as chakra style, is #0060df) and style such that log group appears like a link
   const logGroupStyle = "color:#0060df;cursor:pointer;font-weight:bold;";
 
+  const regExp = /\[(.*?)\] \{(.*?)\} (.*?) -/;
```
Comment (pierrejeambrun) [inline airflow/www/static/js/dag/details/taskInstance/Logs/utils.ts:93]: The regexp was slightly updated, can you elaborate please ?

**Review by pierrejeambrun — COMMENTED** [review]
Tested, working as expected.

Just a few questions.

I think we should also consider applying the colors to those lines. (only certain lines appear red.)

**Comment by potiuk** [comment]
> I noticed that the PR directly targets v2-10-test. I think it's fine in this case.

I think yes, but - we need to make sure this one will be forward-ported (for the lack of better word) or determined not to be needed in main :D

**Comment by jason810496** [comment]
Refactored the test cases to include error cases with traceback.

> I think we should also consider applying the colors to those lines. (only certain lines appear red.)

This has been addressed. Screenshot:
截圖 2024-11-30 下午1 30 11

**Comment by pierrejeambrun** [comment]
Thanks for the update @jason810496 I was really busy today, I'll do a review tomorrow.

**Review by pierrejeambrun — DISMISSED** [review]
Looks good thanks.

Some of the questions of the previous review are left un answered. Just for my comprehension, if do you mind giving me a few hints.

**Review by pierrejeambrun — APPROVED** [review]

**Comment by pierrejeambrun** [comment]
Great thanks @jason810496 for the details.

**Pull request merged**